File: core/src/main/java/com/example/core/utils/TraceLogUtils.java
package com.example.core.utils;

import android.util.Log;

public class TraceLogUtils {

    private static final String TAG = "...";

    public static void log(String msg) {
        StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
        //本身
        StackTraceElement stackTraceElement = stackTrace[3];
        String fun = stackTraceElement.getMethodName();
        String className = stackTraceElement.getFileName().split("\\.")[0];

        //上一层
        StackTraceElement stackTraceElementParent;
        String funParent = "", classNameParent = "";
        for (int i = 4; i < stackTrace.length; i++) {
            stackTraceElementParent = stackTrace[i];
            if (stackTraceElementParent.getMethodName().contains("$")) {
                continue;
            }
            funParent = stackTraceElementParent.getMethodName();
            classNameParent = stackTraceElementParent.getFileName().split("\\.")[0];
            break;
        }


        Log.d(TAG, classNameParent + "." + funParent + ": " + className + " -> " + fun + ": " + msg);
    }

    public static void progressLog(String tag, String msg) {
        StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
        //本身
        StackTraceElement stackTraceElement = stackTrace[3];
        String fun = stackTraceElement.getMethodName();
        String className = stackTraceElement.getFileName().split("\\.")[0];

        //上一层
        StackTraceElement stackTraceElementParent;
        String funParent = "", classNameParent = "";
        for (int i = 4; i < stackTrace.length; i++) {
            stackTraceElementParent = stackTrace[i];
            if (stackTraceElementParent.getMethodName().contains("$")) {
                continue;
            }
            funParent = stackTraceElementParent.getMethodName();
            classNameParent = stackTraceElementParent.getFileName().split("\\.")[0];
            break;
        }

        Log.d(tag, classNameParent + "." + funParent + ": " + className + " -> " + fun + ": " + msg);
    }

}

File: core/src/main/java/com/example/core/utils/TraceLogUtils.kt
package com.example.core.utils

import android.util.Log

fun commonLog(msg: String) {
    Log.d("...", msg)
}

object TraceLogUtils {
    private const val TAG = "..."

    fun log(msg: String) {
        val stackTrace = Thread.currentThread().stackTrace
        //本身
        val stackTraceElement = stackTrace[3]
        val `fun` = stackTraceElement.methodName
        val className =
            stackTraceElement.fileName.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }
                .toTypedArray()[0]

        //上一层
        var stackTraceElementParent: StackTraceElement
        var funParent = ""
        var classNameParent = ""
        for (i in 4 until stackTrace.size) {
            stackTraceElementParent = stackTrace[i]
            if (stackTraceElementParent.methodName.contains("$")) {
                continue
            }
            funParent = stackTraceElementParent.methodName
            classNameParent = stackTraceElementParent.fileName.split(".")[0]
            break
        }


        Log.d(TAG, "$classNameParent.$funParent: $className -> $`fun`: $msg")
    }

    fun progressLog(tag: String?, msg: String) {
        val stackTrace = Thread.currentThread().stackTrace
        //本身
        val stackTraceElement = stackTrace[3]
        val `fun` = stackTraceElement.methodName
        val className =
            stackTraceElement.fileName.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }
                .toTypedArray()[0]

        //上一层
        var stackTraceElementParent: StackTraceElement
        var funParent = ""
        var classNameParent = ""
        for (i in 4 until stackTrace.size) {
            stackTraceElementParent = stackTrace[i]
            if (stackTraceElementParent.methodName.contains("$")) {
                continue
            }
            funParent = stackTraceElementParent.methodName
            classNameParent = stackTraceElementParent.fileName.split(".")[0]
            break
        }

        Log.d(tag, "$classNameParent.$funParent: $className -> $`fun`: $msg")
    }
}
